File: food_order_app/lib/data/order/models/order_model.dart
import '../../../domain/order/entities/order_entity.dart';

class OrderItemModel {
  final String id;
  final String foodId;
  final String foodName;
  final String foodImageUrl;
  final double price;
  final int quantity;

  const OrderItemModel({
    required this.id,
    required this.foodId,
    required this.foodName,
    required this.foodImageUrl,
    required this.price,
    required this.quantity,
  });

  factory OrderItemModel.fromJson(Map<String, dynamic> json) {
    String foodId = '';
    String foodName = 'Food Item';
    String foodImageUrl = '';
    double itemPrice = (json['price'] as num?)?.toDouble() ?? 0.0;

    final foodData = json['food'];
    if (foodData is Map<String, dynamic>) {
      foodId = foodData['_id'] as String? ?? foodData['id'] as String? ?? '';
      foodName = foodData['name'] as String? ?? 'Food Item';
      foodImageUrl = foodData['imageUrl'] as String? ?? '';
      if (itemPrice == 0.0 && foodData['price'] != null) {
        itemPrice = (foodData['price'] as num).toDouble();
      }
    } else if (foodData is String) {
      foodId = foodData;
    }

    return OrderItemModel(
      id: json['_id'] as String? ?? json['id'] as String? ?? '',
      foodId: foodId,
      foodName: foodName,
      foodImageUrl: foodImageUrl,
      price: itemPrice,
      quantity: (json['quantity'] as num?)?.toInt() ?? 1,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    '_id': id,
    'food': {
      '_id': foodId,
      'name': foodName,
      'imageUrl': foodImageUrl,
      'price': price,
    },
    'price': price,
    'quantity': quantity,
  };

  OrderItemEntity toEntity() {
    return OrderItemEntity(
      id: id,
      foodId: foodId,
      foodName: foodName,
      foodImageUrl: foodImageUrl,
      price: price,
      quantity: quantity,
    );
  }
}

class OrderModel {
  final String id;
  final String userId;
  final List<OrderItemModel> items;
  final double totalAmount;
  final String deliveryAddress;
  final String status;
  final String paymentMethod;
  final String paymentStatus;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const OrderModel({
    required this.id,
    required this.userId,
    required this.items,
    required this.totalAmount,
    required this.deliveryAddress,
    required this.status,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.createdAt,
    this.updatedAt,
  });

  factory OrderModel.fromJson(Map<String, dynamic> json) {
    String userId = '';
    final userData = json['user'];
    if (userData is Map<String, dynamic>) {
      userId = userData['_id'] as String? ?? userData['id'] as String? ?? '';
    } else if (userData is String) {
      userId = userData;
    }

    final rawItems = json['items'] as List<dynamic>? ?? [];
    final items = rawItems
        .whereType<Map<String, dynamic>>()
        .map((item) => OrderItemModel.fromJson(item))
        .toList();

    return OrderModel(
      id: json['_id'] as String? ?? json['id'] as String? ?? '',
      userId: userId,
      items: items,
      totalAmount: (json['totalAmount'] as num?)?.toDouble() ?? 0.0,
      deliveryAddress: json['deliveryAddress'] as String? ?? '',
      status: json['status'] as String? ?? 'pending',
      paymentMethod: json['paymentMethod'] as String? ?? 'cash',
      paymentStatus: json['paymentStatus'] as String? ?? 'pending',
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    '_id': id,
    'user': userId,
    'items': items.map((i) => i.toJson()).toList(),
    'totalAmount': totalAmount,
    'deliveryAddress': deliveryAddress,
    'status': status,
    'paymentMethod': paymentMethod,
    'paymentStatus': paymentStatus,
    'createdAt': createdAt.toIso8601String(),
    'updatedAt': updatedAt?.toIso8601String(),
  };

  OrderEntity toEntity() {
    return OrderEntity(
      id: id,
      userId: userId,
      items: items.map((i) => i.toEntity()).toList(),
      totalAmount: totalAmount,
      deliveryAddress: deliveryAddress,
      status: status,
      paymentMethod: paymentMethod,
      paymentStatus: paymentStatus,
      createdAt: createdAt,
      updatedAt: updatedAt,
    );
  }
}

File: food_order_web/src/data/orders/models/order_model.js
import { OrderEntity, OrderItemEntity, OrderStatus } from '../../../domain/orders/entities/order_entity';

export class OrderModel {

  static fromJson(raw = {}) {
    const id = raw._id || raw.id || '';
    
    // Normalize customer user info
    let customerName = 'Guest Customer';
    let customerPhone = '';
    let userId = '';

    if (raw.user && typeof raw.user === 'object') {
      userId = raw.user._id || raw.user.id || '';
      customerName = raw.user.name || raw.user.fullName || raw.user.email || 'Customer';
      customerPhone = raw.user.phone || '';
    } else if (typeof raw.user === 'string') {
      userId = raw.user;
    }

    if (raw.customerName) customerName = raw.customerName;
    if (raw.customerPhone) customerPhone = raw.customerPhone;

    // Normalize items
    const rawItems = Array.isArray(raw.items) ? raw.items : [];
    const items = rawItems.map((item, idx) => {
      const itemId = item._id || item.id || `item-${idx}`;
      let foodId = '';
      let foodName = 'Food Item';
      let foodImageUrl = '';
      let price = Number(item.price) || 0;
      const quantity = Number(item.quantity) || 1;

      if (item.food && typeof item.food === 'object') {
        foodId = item.food._id || item.food.id || '';
        foodName = item.food.name || item.food.title || foodName;
        foodImageUrl = item.food.imageUrl || item.food.image || '';
        if (!price && item.food.price) price = Number(item.food.price);
      } else if (item.food && typeof item.food === 'string') {
        foodId = item.food;
      }

      if (item.foodName) foodName = item.foodName;
      if (item.foodImageUrl) foodImageUrl = item.foodImageUrl;

      return new OrderItemEntity({
        id: itemId,
        foodId,
        foodName,
        foodImageUrl,
        price,
        quantity,
      });
    });

    return new OrderEntity({
      id,
      orderNumber: raw.orderNumber || (id ? `#${id.slice(-6).toUpperCase()}` : '#ORD-0000'),
      userId,
      customerName,
      customerPhone,
      items,
      totalAmount: Number(raw.totalAmount) || 0,
      deliveryAddress:
        typeof raw.deliveryAddress === 'string'
          ? raw.deliveryAddress
          : raw.deliveryAddress?.address || 'Phnom Penh, Cambodia',
      status: (raw.status || OrderStatus.PENDING).toLowerCase(),
      paymentMethod: raw.paymentMethod || 'cash',
      paymentStatus: raw.paymentStatus || 'pending',
      createdAt: raw.createdAt ? new Date(raw.createdAt) : new Date(),
      updatedAt: raw.updatedAt ? new Date(raw.updatedAt) : new Date(),
      notes: raw.notes || raw.specialInstructions || '',
    });
  }

  static toJson(entity) {
    return {
      id: entity.id,
      status: entity.status,
      totalAmount: entity.totalAmount,
      items: entity.items.map((i) => ({
        food: i.foodId,
        foodName: i.foodName,
        foodImageUrl: i.foodImageUrl,
        quantity: i.quantity,
        price: i.price,
      })),
      deliveryAddress: entity.deliveryAddress,
      paymentMethod: entity.paymentMethod,
      paymentStatus: entity.paymentStatus,
    };
  }
}
